### PR TITLE
Remove legacy Span hacks from compiler

### PR DESCRIPTION
**Pull request by agocke**

### CODE REVIEW

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 1]
```diff
@@ -703,27 +703,6 @@ public static int Count<T>(this ImmutableArray<T> items, Func<T, bool> predicate
...
-        [StructLayout(LayoutKind.Sequential)]
-        private struct ImmutableArrayProxy<T>
-        {
-            internal T[] MutableArray;
-        }
-
-        // TODO(https://github.com/dotnet/corefx/issues/34126): Remove when System.Collections.Immutable
-        // provides a Span API
-        internal static T[] DangerousGetUnderlyingArray<T>(this ImmutableArray<T> array)
-            => Unsafe.As<ImmutableArray<T>, ImmutableArrayProxy<T>>(ref array).MutableArray;
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:715]: 📝 This conflicts with #54007

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 2]
```diff
@@ -703,27 +703,6 @@ public static int Count<T>(this ImmutableArray<T> items, Func<T, bool> predicate
...
-
-        internal static ReadOnlySpan<T> AsSpan<T>(this ImmutableArray<T> array)
-            => array.DangerousGetUnderlyingArray();
-
-        internal static ImmutableArray<T> DangerousCreateFromUnderlyingArray<T>([MaybeNull] ref T[] array)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:720]: 📝 This conflicts with #54007

**Comment by jaredpar** [comment]
This PR seems to be mis-titled. It's not removing the hacks, it's migrating them to a different layer of the code base (and removing the tracking comment to delete them).

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs (hunk 2 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:720]: 📝 Based on experimental results I'm leaning towards not merging #54007, so perhaps not an issue.

**Comment by agocke** [comment]
@jaredpar

>  from compiler 😄 

Happy to keep the comment around, although this hack was specifically added for Span and seemed to have grown (in my opinion) undesirable unrelated uses since then. I think that the Workspaces usage is unrelated to the Span API so the comment would have to be different. Suggestions?

**Comment by agocke** [comment]
Took a closer look at the implementation here, found out it was broken and the other API should never have been used

File: src/Compilers/Core/CodeAnalysisTest/Emit/EmitBaselineTests.cs [hunk 3]
```diff
@@ -16,25 +16,28 @@ namespace Microsoft.CodeAnalysis.UnitTests.Emit
...
-            Assert.Throws<ArgumentNullException>(() => EmitBaseline.CreateInitialBaseline(null, debugInfoProvider, localSigProvider, true));
-            Assert.Throws<ArgumentNullException>(() => EmitBaseline.CreateInitialBaseline(peModule, null, localSigProvider, true));
-            Assert.Throws<ArgumentNullException>(() => EmitBaseline.CreateInitialBaseline(mdModule, debugInfoProvider, null, true));
-            Assert.NotNull(EmitBaseline.CreateInitialBaseline(mdModule, debugInfoProvider, localSigProvider, true));
+                Assert.Throws<ArgumentNullException>(() => EmitBaseline.CreateInitialBaseline(null, debugInfoProvider, localSigProvider, true));
+                Assert.Throws<ArgumentNullException>(() => EmitBaseline.CreateInitialBaseline(peModule, null, localSigProvider, true));
+                Assert.Throws<ArgumentNullException>(() => EmitBaseline.CreateInitialBaseline(mdModule, debugInfoProvider, null, true));
+                Assert.NotNull(EmitBaseline.CreateInitialBaseline(mdModule, debugInfoProvider, localSigProvider, true));
+
+            }
```
Comment (jcouv) [inline src/Compilers/Core/CodeAnalysisTest/Emit/EmitBaselineTests.cs:40]: nit: empty line above

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 5)

**Comment by jcouv** [comment]
@jaredpar @dotnet/roslyn-compiler for second review.

**Review by jaredpar — APPROVED** [review]

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SerializableBytes.cs [hunk 4]
```diff
@@ -236,8 +240,23 @@ public byte[] ToArray()
...
+                var chunkIndex = 0;
+                while (count > 0)
+                {
+                    var chunk = chunks[chunkIndex];
+                    var copyCount = Math.Min(chunk.Length, count);
+
+                    builder.AddRange(chunk, copyCount);
+                    count -= copyCount;
+                    chunkIndex++;
+                }
```
Comment (jcouv) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SerializableBytes.cs:256]:
```suggestion
                }

```

**Pull request merged**